compiler/rustc_typeck/src/check/method/suggest.rs

```diff
@@ -391,28 +396,117 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                                    arg.ty(),
+                                                    ConstVariableOrigin {
+                                                        span: rustc_span::DUMMY_SP,
+                                                        kind: ConstVariableOriginKind::MiscVariable,
+                                                    },
+                                                )
+                                                .into(),
+                                            }
+                                        } else {
+                                            arg
+                                        }
+                                    })
+                                    .collect::<Vec<_>>();
+
+                                tcx.value_path_str_with_substs(
+                                    def_actual.did(),
+                                    tcx.intern_substs(&infer_substs),
+                                )
+                            }
+                            _ => self.ty_to_value_string(ty.peel_refs()),
+                        }
+                    } else {
+                        self.ty_to_value_string(actual.peel_refs())
+                    };
+                    if let SelfSource::MethodCall(_) = source {
+                        let first_arg = if let Some(CandidateSource::Impl(impl_did)) = static_sources.get(0) &&
+                            let Some(assoc) = self.associated_value(*impl_did, item_name) {
+                            let sig = self.tcx.fn_sig(assoc.def_id);
+                            if let Some(first) = sig.inputs().skip_binder().get(0) {
+                                if first.peel_refs() == rcvr_ty.peel_refs() {
+                                    None
+                                } else {
+                                    Some(if first.is_region_ptr() {
+                                        if first.is_mutable_ptr() { "&mut " } else { "&" }
+                                    } else {
+                                        ""
+                                    })
                                 }
-                                _ => self.ty_to_value_string(ty.peel_refs()),
+                            } else {
+                                None
                             }
                         } else {
-                            self.ty_to_value_string(actual.peel_refs())
+                            None
+                        };
+                        let mut applicability = Applicability::MachineApplicable;
+                        let args = if let Some((receiver, args)) = args {
+                            // The first arg is the same kind as the receiver
+                            let it = if first_arg.is_some() {
+                                Box::new(std::iter::once(receiver).chain(args.iter()))
+                                    as Box<dyn Iterator<Item = &Expr<'_>>>
```

Comment:
Can you just do code duplication here? Or collect into a `Vec<_>`? Coercing into a boxed iterator here is kinda confusing and this is diagnostics code, so it doesn't need to be efficient necessarily.